File: exercises/kth_largest_in_array_215.java
// Original Solution
class Solution {
    public int findKthLargest(int[] nums, int k) {
      	PriorityQueue<Integer> minHeap = new PriorityQueue<>(k);  
		for (int i = 0; i < k; i++) {
			minHeap.add(nums[i]);	
		}
		for (int i = k; i < nums.length; i++) {
			if (nums[i] <= minHeap.peek()) continue;	
			minHeap.poll();
			minHeap.add(nums[i]);
		}
		return minHeap.peek();
    }
}
// Cheating method exploiting the input constrains and using an array instead of a priority queue
//class Solution {
//    public int findKthLargest(int[] nums, int k) {
//        int[] count = new int[20002];
//        for(int i =0;i<nums.length;i++) {
//            count[nums[i]+10000]++;
//        }
//        for (int i = count.length-1;i>=0;i--) {
//            k = k - count[i];
//            if (k <= 0) return i-10000;
//        }
//        return 0;
//    }
//}

File: exercises/kth_largest_in_array_215.py
# MinHeap Solution
from heapq import heapreplace, heapify
from typing import List
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # Min heap approach - O(n log k)
        heap = nums[:k]
        heapify(heap)  # O(k)
        
        for num in nums[k:]:
            if num > heap[0]:
                heapreplace(heap, num)
        
        return heap[0]

# Array Solutions
# class Solution:
#     def findKthLargest(self, nums: List[int], k: int) -> int:
#         # Your counting sort approach
#         count = [0] * 20002  # Using list instead of numpy
#         for num in nums:
#             count[num + 10000] += 1
#         for i in range(len(count) - 1, -1, -1):
#             k -= count[i]
#             if k <= 0:
#                 return i - 10000
#         return 0
